### PR TITLE
memcached cache: switch to AWS elasticache-java-cluster-client and add TLS support 

### PR DESCRIPTION
**Pull request by pagrawal10**
### Description

This PR updates the library used for Memcached client to AWS Elasticache Client : https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-java

<b>Why?
</b>
This enables us to use the option of encrypting data in transit:
[Amazon ElastiCache for Memcached now supports encryption of data in transit](https://www.amazonaws.cn/en/new/2022/amazon-elasticache-memcached-supports-encryption-data-in-transit/) 

For clusters running the Memcached engine, ElastiCache supports Auto Discovery—the ability for client programs to automatically identify all of the nodes in a cache cluster, and to initiate and maintain connections to all of these nodes.
[Benefits of Auto Discovery - Amazon ElastiCache](https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/AutoDiscovery.Benefits.html) 

<br>
<b>What?
</b>

AWS has forked spymemcached 2.12.1, and has since added all the patches included in 2.12.2 and 2.12.3 as part of the 1.2.0 release. So, this can now be considered as an equivalent drop-in replacement.

[GitHub - awslabs/aws-elasticache-cluster-client-memcached-for-java: Amazon ElastiCache Cluster Client for Java - enhanced library to connect to ElastiCache clusters.](https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-java) 
https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/elasticache/AmazonElastiCacheClient.html#AmazonElastiCacheClient--
<br>

<b>How?
</b>
How to enable TLS with Elasticache
- On server side:
https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/in-transit-encryption-mc.html#in-transit-encryption-enable-existing-mc

- On client side:
[GitHub - awslabs/aws-elasticache-cluster-client-memcached-for-java: Amazon ElastiCache Cluster Client for Java - enhanced library to connect to ElastiCache clusters.](https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-java#configuring-the-client-in-tls-mode) 

 

How to leverage autodiscovery with AWS Elasticache
[Using Auto Discovery - Amazon ElastiCache](https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/AutoDiscovery.Using.html) 
[Using the ElastiCache Cluster Client for Java - Amazon ElastiCache](https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/AutoDiscovery.Using.ModifyApp.Java.html)

#### Release note

Using AWS Elasticache Client instead of net.spy.memcached client. Allows us to encrypt data in transit. 

<hr>

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by xvrl** [comment]
@pagrawal10 Can we add some reasoning to the PR explain why the client is an acceptable drop-in replacement for the existing one and what we can expect in terms of ongoing maintenance? It might be good to mention that the AWS version is a fork of the existing one.

Also, since you mention security, do you mind updating the documentation to cover:
- how to enable TLS with Elasticache
- how to leverage autodiscovery with AWS Elasticache

File: pom.xml [hunk 1]
```diff
@@ -766,9 +766,9 @@
                 <version>3.3.6</version>
             </dependency>
             <dependency>
-                <groupId>net.spy</groupId>
-                <artifactId>spymemcached</artifactId>
-                <version>2.12.3</version>
+                <groupId>com.amazonaws</groupId>
+                <artifactId>elasticache-java-cluster-client</artifactId>
+                <version>1.2.0</version>
```
Comment (xvrl) [inline pom.xml:771]: it might be worth mentioning in the PR description that AWS forked spymemcached 2.12.1, and has since added all the patches included in 2.12.2 and 2.12.3 as part of the 1.2.0 release, meaning this can now be considered an equivalent drop-in replacement.

File: server/src/main/java/org/apache/druid/client/cache/MemcachedCacheConfig.java [hunk 2]
```diff
@@ -63,6 +63,12 @@ public class MemcachedCacheConfig
   @JsonProperty
   private String locator = "consistent";
 
+  @JsonProperty
+  private boolean sslconnection = false;
```
Comment (xvrl) [inline server/src/main/java/org/apache/druid/client/cache/MemcachedCacheConfig.java:67]: `TLS` is the more correct name for this feature, see https://github.com/memcached/memcached/wiki/TLS and https://github.com/awslabs/aws-elasticache-cluster-client-memcached-for-java#configuring-the-client-in-tls-mode

I would suggest using `enableTls` for this config, which also aligns better with other Druid configs (e.g. https://druid.apache.org/docs/latest/operations/tls-support/#general-configuration)

File: server/src/main/java/org/apache/druid/client/cache/MemcachedCacheConfig.java [hunk 3]
```diff
@@ -63,6 +63,12 @@ public class MemcachedCacheConfig
...
+
+  @JsonProperty
+  private boolean autodiscovery = false;
```
Comment (xvrl) [inline server/src/main/java/org/apache/druid/client/cache/MemcachedCacheConfig.java:70]: let's use `autoDiscovery` here to keep in line with existing camel-case naming conventions used for other configs

File: server/src/main/java/org/apache/druid/client/cache/MemcachedCache.java [hunk 4]
```diff
@@ -355,9 +364,23 @@ public void updateHistogram(String name, int amount)
...
+        sslContext.init(null, tmf.getTrustManagers(), null);
+        // Create the client in TLS mode
+        connectionFactoryBuilder.setSSLContext(sslContext);
+      }
+      if(config.autoDiscovery()) {
+        connectionFactoryBuilder.setClientMode(ClientMode.Dynamic);
+      }
+      else {
+        connectionFactoryBuilder.setClientMode(ClientMode.Static);
+      }
```
Comment (xvrl) [inline server/src/main/java/org/apache/druid/client/cache/MemcachedCache.java:382]: maybe we should call the config `clientMode` = `static` / `dynamic` here to better align with AWS naming rather than using our own

File: docs/configuration/index.md [hunk 5]
```diff
@@ -2103,18 +2103,18 @@ In addition to the normal cache metrics, the caffeine cache implementation also
...
+| `druid.cache.maxObjectSize`   | Maximum object size in bytes for a Memcached object.                                                   | 52428800 (50 MiB) |
+| `druid.cache.memcachedPrefix` | Key prefix for all keys in Memcached.                                                                  | druid             |
+| `druid.cache.numConnections`  | Number of memcached connections to use.                                                                | 1                 |
+| `druid.cache.protocol`        | Memcached communication protocol. Can be binary or text.                                               | binary            |
+| `druid.cache.locator`         | Memcached locator. Can be consistent or array_mod.                                                     | consistent        |
+| `druid.cache.enableTls`       | Use TLS based ssl connection for Memcached client.  Boolean                                            | false             |
+| `druid.cache.clientMode`      | Client Mode. Dynamic mode uses AutoDiscovery feature of AWS Memcached. String. "static" or "dynamic"   | static            |
```
Comment (xvrl) [inline docs/configuration/index.md:2117]: two things:
- can we explain that "static" means individual cluster nodes have to be configured?
- can we add a link to the AWS docs to mention how to configure the `druid.cache.hosts` when using dynamic mode?

File: docs/configuration/index.md [hunk 6]
```diff
@@ -2103,18 +2103,18 @@ In addition to the normal cache metrics, the caffeine cache implementation also
...
-| `druid.cache.enableTls`       | Use TLS based ssl connection for Memcached client.  Boolean                                                    | false             |
-| `druid.cache.clientMode`      | ClientMode. Dynamic mode uses AutoDiscovery feature of AWS Elasticache Memcache. String. "static" or "dynamic" | static            |
+| Property                      | Description                                                                                            | Default           |
+|-------------------------------|--------------------------------------------------------------------------------------------------------|-------------------|
+| `druid.cache.expiration`      | Memcached [expiration time](https://code.google.com/p/memcached/wiki/NewCommands#Standard_Protocol).   | 2592000 (30 days) |
+| `druid.cache.timeout`         | Maximum time in milliseconds to wait for a response from Memcached.                                    | 500               |
+| `druid.cache.hosts`           | Comma separated list of Memcached hosts `<host:port>`.                                                 | none              |
```
Comment (xvrl) [inline docs/configuration/index.md:2110]: can we update the documentation here to explain that list of hosts applies for `static` mode and how that changes when clientMode is set to `dynamic`?

File: server/src/test/java/org/apache/druid/client/cache/MemcachedCacheTest.java [hunk 7]
```diff
@@ -224,6 +217,26 @@ public void emit(Event event)
...
+        return true;
+      }
+
+      @Override
+      public String getHosts()
+      {
+        return "localhost:9999";
+      }
+    };
+    MemcachedCache.create(config);
```
Comment (xvrl) [inline server/src/test/java/org/apache/druid/client/cache/MemcachedCacheTest.java:237]: this doesn't seem to test much other than making sure setting enableTls doesn't error out. Can we at least have the test somehow verify that the client has tls enabled?

File: docs/configuration/index.md [hunk 8]
```diff
@@ -2103,16 +2103,18 @@ In addition to the normal cache metrics, the caffeine cache implementation also
...
+| `druid.cache.timeout`         | Maximum time in milliseconds to wait for a response from Memcached.                                                                                                                                                                                                                                                                                                                                                                                                             | 500               |
+| `druid.cache.hosts`           | Comma separated list of Memcached hosts `<host:port>`. Needed only when druid.cache.clientMode=static. Dynamic mode [automatically identifies nodes in your cluster](https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/AutoDiscovery.html)                                                                                                                                                                                                                            | none              |
```
Comment (xvrl) [inline docs/configuration/index.md:2110]: are we sure this is not needed in dynamic mode? How does one specify which cluster to connect to?

File: docs/configuration/index.md [hunk 9]
```diff
@@ -2103,16 +2103,18 @@ In addition to the normal cache metrics, the caffeine cache implementation also
...
+| `druid.cache.locator`         | Memcached locator. Can be consistent or array_mod.                                                                                                                                                                                                                                                                                                                                                                                                                              | consistent        |
+| `druid.cache.enableTls`       | Use TLS based ssl connection for Memcached client.  Boolean                                                                                                                                                                                                                                                                                                                                                                                                                     | false             |
```
Comment (xvrl) [inline docs/configuration/index.md:2116]: nit, ssl is redundant here
```suggestion
| `druid.cache.enableTls`       | enable TLS based connection for Memcached client.  Boolean                                                                                                                                                                                                                                                                                                                                                                                                                     | false             |
```

File: server/src/main/java/org/apache/druid/client/cache/MemcachedCache.java [hunk 10]
```diff
@@ -355,9 +360,23 @@ public void updateHistogram(String name, int amount)
...
+      if(config.enableTls()) {
+        // Build SSLContext
+        TrustManagerFactory tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+        tmf.init((KeyStore) null);
+        SSLContext sslContext = SSLContext.getInstance("TLS");
+        sslContext.init(null, tmf.getTrustManagers(), null);
+        // Create the client in TLS mode
+        connectionFactoryBuilder.setSSLContext(sslContext);
+      }
+      if(Objects.equals(config.getClientMode(), "dynamic")) {
```
Comment (xvrl) [inline server/src/main/java/org/apache/druid/client/cache/MemcachedCache.java:373]: nit, Objects.equals seems overkill here, why not simply write `"dynamic".equals(config.getClientMode())`

File: server/src/main/java/org/apache/druid/client/cache/MemcachedCache.java [hunk 11]
```diff
@@ -355,9 +360,23 @@ public void updateHistogram(String name, int amount)
...
+        connectionFactoryBuilder.setClientMode(ClientMode.Dynamic);
+      }
+      else {
+        connectionFactoryBuilder.setClientMode(ClientMode.Static);
```
Comment (xvrl) [inline server/src/main/java/org/apache/druid/client/cache/MemcachedCache.java:377]: we should throw an error if someone provides an invalid client mode rather than defaulting to static

Comment (pagrawal10) [reply server/src/main/java/org/apache/druid/client/cache/MemcachedCache.java:377]: That change will not be backward compatible. As per my understanding, static mode mirrors the functionality present currently with spy.memcache. In case, the users do not provide any mode, we should default to static mode.

File: server/src/test/java/org/apache/druid/client/cache/MemcachedCacheTest.java (hunk 7 above)
Comment (pagrawal10) [reply server/src/test/java/org/apache/druid/client/cache/MemcachedCacheTest.java]: Updated the test

File: server/src/main/java/org/apache/druid/client/cache/MemcachedCache.java (hunk 11 above)
Comment (xvrl) [reply server/src/main/java/org/apache/druid/client/cache/MemcachedCache.java:378]: I'm not saying we should throw an error when client mode is left out, we should still default to static if no client mode is provided. However, if an invalid client mode is provided we should throw an error.

Comment (pagrawal10) [reply server/src/main/java/org/apache/druid/client/cache/MemcachedCache.java:375]: Updated

File: server/src/main/java/org/apache/druid/client/cache/MemcachedCache.java [hunk 12]
```diff
@@ -355,9 +358,28 @@ public void updateHistogram(String name, int amount)
...
+        // Create the client in TLS mode
+        connectionFactoryBuilder.setSSLContext(sslContext);
+      }
+      if ("dynamic".equals(config.getClientMode())) {
+        connectionFactoryBuilder.setClientMode(ClientMode.Dynamic);
+        connectionFactoryBuilder.setHostnameForTlsVerification(config.getHosts().split(",")[0]);
+      } else if ("static".equals(config.getClientMode())) {
+        connectionFactoryBuilder.setClientMode(ClientMode.Static);
+      } else {
+        throw new RuntimeException("Invalid value provided for `druid.cache.clientMode`. Value must be 'static' or 'dynamic'.");
```
Comment (xvrl) [inline server/src/main/java/org/apache/druid/client/cache/MemcachedCache.java:377]: cache configurations can be nested as l1/l2 so it's not always the case that the prefix for this configuration is `druid.cache.`, so I would remove the prefix. We should also add the value to make it clear what the issue was.

```suggestion
        throw new RE("Invalid value provided for memcached client configuration `clientMode` [{}]. Value must be 'static' or 'dynamic'.", config.getClientMode());
```

File: server/src/test/java/org/apache/druid/client/cache/MemcachedCacheTest.java [hunk 13]
```diff
@@ -224,6 +217,37 @@ public void emit(Event event)
...
+      }
+
+      @Override
+      public String getHosts()
+      {
+        return "localhost:9999";
+      }
+    };
+    // Static Mode
+    Assert.assertEquals(config.getClientMode(), "static");
```
Comment (xvrl) [inline server/src/test/java/org/apache/druid/client/cache/MemcachedCacheTest.java:238]: this test seems to ensure that the default client mode is static, which is fine, but it doesn't actually do what the test name suggests, which is to verify that ssl is configured. Can we move this to its own test to verify the default?

File: server/src/test/java/org/apache/druid/client/cache/MemcachedCacheTest.java [hunk 14]
```diff
@@ -224,6 +217,37 @@ public void emit(Event event)
...
+    MemcachedCache client = new MemcachedCache(
+            Suppliers.ofInstance(
+                    StupidResourceHolder.create(new MockMemcachedClient())
+            ),
+            config,
+            NOOP_MONITOR
+    );
+    Assert.assertNull(client.get(new Cache.NamedKey("a", HI)));
+    put(client, "a", HI, 1);
+    Assert.assertEquals(1, get(client, "a", HI));
```
Comment (xvrl) [inline server/src/test/java/org/apache/druid/client/cache/MemcachedCacheTest.java:248]: this code doesn't actually verify anything SSL related, since all the ssl configuration handling is done in `MemcachedCache.create(...)` and the supplier you provide is creating its own mock instances independently.

If we want to ensure the connection factory is actually configured correctly, you will probably have to refactor the create method a little bit to move the connection factory creation into its own method which you can then test independently.

File: website/static/css/custom.css [hunk 15]
```diff
@@ -99,15 +99,3 @@ article iframe {
...
-  font-weight: bold;
-}
-.postAPI {
-  color: #00bf7d; 
-  font-weight: bold;
-}
-.deleteAPI {
-  color: #f49200; 
-  font-weight: bold;
-}
```
Comment (xvrl) [inline website/static/css/custom.css:113]: why are we changing this?

File: server/src/main/java/org/apache/druid/client/cache/MemcachedCustomConnectionFactoryBuilder.java [hunk 16]
```diff
@@ -213,6 +204,45 @@ public long getAuthWaitTime()
...
+                + getHashAlg().toString() + " Max Reconnect Delay: "
+                + getMaxReconnectDelay() + ", Max Op Timeout: " + getOperationTimeout()
+                + ", Op Queue Length: " + getOpQueueLen() + ", Op Max Queue Block Time"
+                + getOpQueueMaxBlockTime() + ", Max Timeout Exception Threshold: "
+                + getTimeoutExceptionThreshold() + ", Read Buffer Size: "
+                + getReadBufSize() + ", Transcoder: " + getDefaultTranscoder()
+                + ", Operation Factory: " + getOperationFactory() + " isDaemon: "
+                + isDaemon() + ", Optimized: " + shouldOptimize() + ", Using Nagle: "
+                + useNagleAlgorithm() + ", KeepAlive: " + getKeepAlive() + ", SSLContext: " + getSSLContext() + ", ConnectionFactory: " + getName();
+      }
```
Comment (xvrl) [inline server/src/main/java/org/apache/druid/client/cache/MemcachedCustomConnectionFactoryBuilder.java:245]: any reason we are overriding this?

File: website/static/css/custom.css [hunk 17]
```diff
@@ -98,4 +98,4 @@ article iframe {
   margin-left: auto;
   margin-right: auto;
   max-width: 100%;
-}
+}
```
Comment (xvrl) [inline website/static/css/custom.css:101]: can we avoid unnecessary changes to this file since we are not changing it?

**Review by xvrl — APPROVED** [review]
LGTM once tests are passing
